### PR TITLE
Rework on substring index

### PR DESCRIPTION
**Pull request by Feng-Jiang28**
Depends on https://github.com/NVIDIA/spark-rapids-jni/pull/2205
This PR is for closing issue https://github.com/NVIDIA/spark-rapids/issues/8750, by replace the regex implementation underneath with `cudf::strings::slice_strings()`  with `cudf::strings::find()/rfind()`, new implementation achieved 12X speed up comparing to the old version https://github.com/NVIDIA/spark-rapids-jni/pull/2205#issuecomment-2225342533

### CODE REVIEW

**Comment by thirtiseven** [comment]
Depends on https://github.com/NVIDIA/spark-rapids-jni/pull/2205

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -1660,17 +1614,14 @@ case class GpuSubstringIndex(strExpr: Expression,
...
-        }
-      }
-    } else {
-      withResource(str.getBase.extractRe(new RegexProgram(regexp))) { table: Table =>
-        table.getColumn(0).incRefCount()
-      }
+
+    val strColumnView = str.getBase
+    val delimiter = delim.getBase
+    val cnt = count.getValue.asInstanceOf[Int]
```
Comment (thirtiseven) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1620]: For GpuScalar, we need to check if count `isValid` before get its value

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala [hunk 2]
```diff
@@ -66,7 +66,7 @@ class RapidsTestSettings extends BackendTestSettings {
   enableSuite[RapidsMathFunctionsSuite]
   enableSuite[RapidsRegexpExpressionsSuite]
   enableSuite[RapidsStringExpressionsSuite]
-    .exclude("string substring_index function", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/8750"))
+    //.exclude("string substring_index function", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/8750"))
```
Comment (thirtiseven) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala:69]: We can just remove this line.

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/suites/RapidsStringExpressionsSuite.scala [hunk 3]
```diff
@@ -19,7 +19,38 @@
...
 package org.apache.spark.sql.rapids.suites
 
-import org.apache.spark.sql.catalyst.expressions.StringExpressionsSuite
+import org.apache.spark.sql.catalyst.expressions.{Literal, StringExpressionsSuite, SubstringIndex}
 import org.apache.spark.sql.rapids.utils.RapidsTestsTrait
+import org.apache.spark.sql.types.StringType
 
-class RapidsStringExpressionsSuite extends StringExpressionsSuite with RapidsTestsTrait {}
+class RapidsStringExpressionsSuite extends StringExpressionsSuite with RapidsTestsTrait {
+  test("string substring_index function in rapids") {
```
Comment (thirtiseven) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/suites/RapidsStringExpressionsSuite.scala:27]: Question: Why do we need this case? Is this modified from a Spark UT case or is it a new added case? If it's new, I think we can add it to our Scala test part.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 4]
```diff
@@ -1660,17 +1614,14 @@ case class GpuSubstringIndex(strExpr: Expression,
...
+    withResource(GpuSubstringIndexUtils.substringIndex(strColumnView, delimiter, cnt)) { result =>
```
Comment (thirtiseven) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1621]: `result` is not an intermediate value so I think we don't need to manage it's resource. Just `GpuSubstringIndexUtils.substringIndex(str.getBase, delim.getBase, cnt)` will be work.

**Review by thirtiseven — COMMENTED** [review]
Better to have some new integration tests for new covered cases so we can test them on all spark versions.

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala (hunk 2 above)
Comment (Feng-Jiang28) [reply tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala]: Removed already. Thank you!

**Comment by Feng-Jiang28** [comment]
build

**Comment by Feng-Jiang28** [comment]
build

**Comment by Feng-Jiang28** [comment]
build

**Comment by Feng-Jiang28** [comment]
build

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/suites/RapidsStringExpressionsSuite.scala [hunk 5]
```diff
@@ -22,4 +22,4 @@ package org.apache.spark.sql.rapids.suites
 import org.apache.spark.sql.catalyst.expressions.StringExpressionsSuite
 import org.apache.spark.sql.rapids.utils.RapidsTestsTrait
 
-class RapidsStringExpressionsSuite extends StringExpressionsSuite with RapidsTestsTrait {}
+class RapidsStringExpressionsSuite extends StringExpressionsSuite with RapidsTestsTrait {}
```
Comment (thirtiseven) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/suites/RapidsStringExpressionsSuite.scala:25]: nit: add a new line after the end of file

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 6]
```diff
@@ -1654,22 +1608,13 @@ case class GpuSubstringIndex(strExpr: Expression,
...
-        }
-      }
-    } else {
-      withResource(str.getBase.extractRe(new RegexProgram(regexp))) { table: Table =>
-        table.getColumn(0).incRefCount()
-      }
+    if(delim.isValid && count.isValid){
+      GpuSubstringIndexUtils.substringIndex(str.getBase,  delim.getBase,
+        count.getValue.asInstanceOf[Int])
+    }else{
```
Comment (thirtiseven) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1616]:
```suggestion
    if (delim.isValid && count.isValid) {
      GpuSubstringIndexUtils.substringIndex(str.getBase, delim.getBase,
          count.getValue.asInstanceOf[Int])
    } else {
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 7]
```diff
@@ -1654,22 +1608,13 @@ case class GpuSubstringIndex(strExpr: Expression,
...
   override def doColumnar(str: GpuColumnVector, delim: GpuScalar,
       count: GpuScalar): ColumnVector = {
-    if (regexp == null) {
-      withResource(str.getBase.isNull) { isNull =>
-        withResource(Scalar.fromString("")) { emptyString =>
-          isNull.ifElse(str.getBase, emptyString)
-        }
-      }
+    if (delim.isValid && count.isValid) {
+      GpuSubstringIndexUtils.substringIndex(str.getBase,  delim.getBase,
```
Comment (thirtiseven) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1614]: nit: 
```suggestion
      GpuSubstringIndexUtils.substringIndex(str.getBase, delim.getBase,
```

**Review by thirtiseven — COMMENTED** [review]
LGTM

**Comment by Feng-Jiang28** [comment]
build

**Review by revans2 — APPROVED** [review]
Performance is much much better and is super fast compared to the CPU.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 8]
```diff
@@ -1584,63 +1585,16 @@ class SubstringIndexMeta(
     override val parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule)
     extends TernaryExprMeta[SubstringIndex](expr, conf, parent, rule) {
-  private var regexp: String = _
 
-  override def tagExprForGpu(): Unit = {
-    val delim = GpuOverrides.extractStringLit(expr.delimExpr).getOrElse("")
-    if (delim == null || delim.length != 1) {
-      willNotWorkOnGpu("only a single character deliminator is supported")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1592]: This is removing the single-character delimiter restriction, but GpuOverrides still has a PS note that states this limitation.  That PS note should be removed and docs regenerated.

**Comment by Feng-Jiang28** [comment]
build

**Comment by Feng-Jiang28** [comment]
build

**Comment by Feng-Jiang28** [comment]
build

**Review by jlowe — COMMENTED** [review]
As I mentioned in my comment, docs need to be regenerated.  Premerge checks are failing because of this.  You can run `mvn verify -DskipTests` to regenerate the docs and then commit the changes.
```
[2024-07-24T13:11:36.608Z] found modified files during mvn verify:
[2024-07-24T13:11:36.609Z]  M docs/supported_ops.md
```

**Review by jlowe — APPROVED** [review]

**Comment by jlowe** [comment]
build

**Pull request merged**